core: add interrupt framework

jenswi-linaro opened the pull request:
Adds interrupt frameworks and adjusts gic driver to fit in.

Update plat-vexpress and sunxi platforms to initialize gic with slightly
modified interface.

Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (QEMU, FVP)
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswi-linaro commented:
The interrupt framework doesn't currently have any locks. That's not a problem for the moment, but if we add handlers after boot time it will become a problem. Should we address this now or wait until the system is more dynamic?

pascal-brand-st-dev commented on core/drivers/gic.c:
```diff
@@ -152,154 +162,205 @@ void gic_init(vaddr_t gicc_base, vaddr_t gicd_base)
+	struct gic_data *gd = container_of(chip, struct gic_data, chip);
+
+	gic_it_add(gd, it);
+	gic_it_set_cpu_mask(gd, it, 0xff);
```
`0xff` used to be `gic_it_get_target(0)` is it an issue to have this hard-coded value?

jenswi-linaro replied:
0xff means any cpu online, while the old evaluated to this cpu only. I'll add a comment.

pascal-brand-st-dev commented on core/arch/arm/plat-vexpress/main.c:
```diff
@@ -66,58 +69,32 @@ static const struct thread_handlers handlers = {
-#elif PLATFORM_FLAVOR_IS(qemu_virt)
 void main_init_gic(void)
 {
+#if PLATFORM_FLAVOR_IS(fvp) || PLATFORM_FLAVOR_IS(juno) || \
```
Can it use CFG_ARM32 instead of platform flavor? It can be part of another PR as not directly related to this one.

jenswi-linaro replied:
FVP and Juno supports both arm32 and arm64.

pascal-brand-st-dev replied:
Right. Then I would prefer to have a new CFG_WITH_ARM_TF instead of this. But this is only an idea to keep in mind for another PR in case it sounds good.

pascal-brand-st-dev commented:
About the lock that is missing, I am ok with it now. However, can you add a "Fixme" comment just to remember.

Also, few travis checkpatch errors. Not sure it make sense to fix.

With or without my comments:
`Reviewed-by: Pascal Brand <pascal.brand@linaro.org>`

jenswi-linaro commented:
Both checkpatch warnings are false positives in my opinion.

I'll add a warning about locking.

pascal-brand-st-dev commented:
`Reviewed-by: Pascal Brand <pascal.brand@linaro.org>`

jenswi-linaro commented:
Rebased and tag applied.
I've added the comments discussed above. I've also made a small change to plat-sunxi to pass the pointer to `gic_data` in for `r0` for `gic_cpu_init`

pascal-brand-st-dev commented:
`Reviewed-by: Pascal Brand <pascal.brand@linaro.org>`

jenswi-linaro commented:
Squashed.
